src/python/pants/util/objects.py

```diff
@@ -52,6 +55,184 @@ def __getnewargs__(self):
+      full_msg, *args, **kwargs)
+
+
+class TypeCheckError(TypedDatatypeInstanceConstructionError):
+
+  def __init__(self, type_name, msg, *args, **kwargs):
+    formatted_msg = "type check error:\n{}".format(msg)
+    super(TypeCheckError, self).__init__(
+      type_name, formatted_msg, *args, **kwargs)
+
+
+class TypeDecl(AbstractClass):
+
+  class ConstructionError(Exception):
+    pass
+
+  @abstractmethod
+  def matches_value(self, val):
+    """Return whether or not the argument matches the type described by this
+  class.
+    """
+
+  @abstractmethod
+  def as_union(self):
+    """Return a version of this object which is an instance of Union."""
+
+  def compose(self, rhs):
+    """Return a TypeDecl which matches either this or another TypeDecl."""
+    self_types = self.as_union().types
+    rhs_types = rhs.as_union().types
+    all_types = self_types + rhs_types
+    return Union(*all_types)
+
+
+class SimpleTypeDecl(datatype('SimpleTypeDecl', ['matching_type']), TypeDecl):
+
+  def __new__(cls, matching_type):
+    if not isinstance(matching_type, type):
+      raise cls.ConstructionError(
+        "argument should be a type: '{}'".format(matching_type))
+
+    return super(SimpleTypeDecl, cls).__new__(cls, matching_type)
+
+  def matches_value(self, val):
+    return isinstance(val, self.matching_type)
+
+  def as_union(self):
+    return Union(self.matching_type)
+
+
+class Union(datatype('Union', ['types']), TypeDecl):
+
+  def __new__(cls, *types):
+    if len(types) == 0:
+      raise cls.ConstructionError("at least one type must be provided to Union")
+    if not all([isinstance(ty, type) for ty in types]):
+      raise cls.ConstructionError(
+        "all arguments to Union should be types: '{}'".format(types))
+
+    return super(Union, cls).__new__(cls, types)
+
+  def matches_value(self, val):
+    return any([isinstance(val, ty) for ty in self.types])
+
+  def as_union(self):
+    return self
+
+
+def typed_datatype(type_name, field_decls):
+  """A wrapper over namedtuple which accepts a dict of field names and types.
+
+  This can be used to very concisely define classes which have fields that are
+  type-checked at construction.
+  """
+
+  type_name = str(type_name)
+
+  if not isinstance(field_decls, dict):
+    raise TypedDatatypeClassConstructionError(
+      type_name,
+      "field_decls is not a dict: '{}'".format(field_decls))
+
+  if not field_decls:
+    raise TypedDatatypeClassConstructionError(
+      type_name,
+      "no fields were declared")
+
+  # Turn every type declaration into an instance of TypeDecl, and place the
+  # results in processed_type_decls.
+  # TODO: Make this kind of exception pattern (filter for errors then display
+  # them all at once) more ergonomic.
+  processed_type_decls = {}
+  invalid_type_decls = []
+  for name, cls in field_decls.items():
+    if isinstance(cls, TypeDecl):
+      processed_type_decls[name] = cls
+      continue
```

Comment:
Rather than `continue; if`, this looks like it would sit pretty naturally as a chain of `elif`s?

Reply:
I'm pretty sure you're right, I was just having trouble reading the `elif` version in my head when I wrote this bit, so used `continue`. I will make this change.

Reply:
I converted this into [`Exactly.from_type_or_collection()`](https://github.com/pantsbuild/pants/blob/ae70e028fd14a12ab3fbe10f9ceb0c92eea6027c/src/python/pants/util/objects.py#L193) when I [moved `TypeConstraint` into `objects.py`](https://github.com/pantsbuild/pants/commit/ae70e028fd14a12ab3fbe10f9ceb0c92eea6027c). This actually has the same control flow as the original, but `return` instead of `continue` is a little bit more sane (imo) -- let me know if I'm wrong about that.